### ISSUE TITLE
best practices for keeping parent to children tree in sync

### ISSUE THREAD

**Issue by chino23** [pouchdb/pouchdb#3652]
Let's say I have the following setup, each of those is a document:
````
Project
    Category
        Task 1
        Task 2
             Subtask
````

So the subtask has a reference to the id of the Task 2, Task 1 and 2 have a reference to the id of Category, and Category has a reference to the id of Project. That way it can be stored as individual documents in the database and then I can recreate the tree in the app by following the references from the back.

Let's assume the following:

User A and B have synced the above tree and it's the same on their system.

Now user A deletes the Project, and syncs.

User B is offline now, and adds another Subtask to Task 2. Then later when he gets online, he syncs. 

Now what happens is that user B gets synced the deleted items of the whole tree, making it disappear. But he also ends up with an orphaned Subtask referencing an id of Task 2 that doesn't exist anymore.

What's the best practice here to avoid that situation? I'm thinking that one way would be to update the rev of every parent element when adding a child, so at least when he adds it later, these later revisions will win. 

The only problem with that approach is when someone spams the 'add task' button, a lot of parent updates will pile up at once.

Any thoughts? I'd highly appreciate it!

**Comment by nolanlawson** [comment]
The smallest unit of atomicity in CouchDB/PouchDB is a single document, so in general it's recommended to stuff everything inside of a single document. Otherwise your database can end up in an illogical state, like what you describe.

To be fair, you can use tricks like the relational-pouch plugin or linked documents to simulate a relational schema (with ad-hoc relations between the different documents), but then you will need to do more work to resolve conflicts in cases like this. E.g. you will need to decide what to do with orphans and resolve that yourself. Also consider the case where the parent is in one conflicted state while the child is in another conflicted state if you really want to make your head spin. :)

So basically:

1) stuff everything in a single document -> your CRUD logic is harder, but your conflict-resolution logic is easier
2) create ad-hoc relations -> your CRUD logic is easier, but your conflict-resolution logic is harder

**Comment by chino23** [comment]
Thanks a lot for the answer @nolanlawson, highly appreciate it.

I've read an article about how to best set up the documents for efficient syncing. It was talking about putting each element into a single document, and then referencing it. While it keeps the docs small, I see the problem with the orphaned elements and conflict resolution that can get messy

But here's the thing: Imagine I put all the things into one document, like the whole tree above. Wouldn't that mean that

1) The doc will get pretty big and takes long time to sync
2) What happens if User A updates a lot, adds, edits, renames etc. and a second later user B does similar things. How is this supposed to being tracked which elements to update inside a document? I know it would take some effort to figure that out but to me it just seems so... off in the approach.  Am I wrong?

I'd love to know if there's some best case practice I could have a look at, and I'm sure a lot of others would find that very insightful as it's a very basic sync pattern.

**Comment by nolanlawson** [comment]
It's a common problem designing CouchDB databases. You might have more luck if you ask around in the #couchdb room on Freenode IRC, where I'm sure they've seen every possible schema under the sun.

One common approach so that you never have to deal with conflicts is the "every doc is a delta" approach, described here: http://pouchdb.com/guides/conflicts.html

But in general, yeah, the problem with dividing up your documents vs combining them together implies exactly the tradeoff you describe. There really is no right answer, just whichever one you find easier to manage.

**Comment by chino23** [comment]
Thanks Nolan, I'll have a look at that and talk to the couchdb guys. I'll post some updates here then.
